Upload to nyaa.si, anidex & TokyoTosho [done]

Kiloutre opened the pull request:
#1599

Kiloutre committed:
Update upload.go

Kiloutre committed:
Update helpers.go

Kiloutre committed:
Update template_test.go

Kiloutre committed:
Update upload_multiple.jet.html

Kiloutre committed:
Merge branch 'dev' into upload-multiple

Kiloutre committed:
Merge branch 'dev' into upload-multiple

Kiloutre committed:
change variable names

Kiloutre committed:
wrong copypaste

Kiloutre committed:
change variable name

Kiloutre committed:
ditto

Kiloutre committed:
Fix travis

Kiloutre committed:
Update main.css

Kiloutre committed:
Update upload.go

Kiloutre committed:
Update upload.jet.html

Kiloutre committed:
Update main.css

Kiloutre committed:
Update upload.go

Kiloutre committed:
More compact form

Kiloutre committed:
CSS changes to go along

Kiloutre committed:
Update main.css

Kiloutre committed:
Update upload.jet.html

Kiloutre committed:
Slightly lower bottom margin

Kiloutre committed:
Update upload_multiple.jet.html

Kiloutre committed:
Merge branch 'dev' into upload-multiple

Kiloutre committed:
tomorrow color adjustements

Kiloutre committed:
small css adjustements

Kiloutre committed:
Update upload.go

Kiloutre committed:
Update default_config.yml

Kiloutre committed:
Update default_config.yml

Kiloutre committed:
Update structs.go

Kiloutre committed:
Update user.go

akuma06 commented on config/structs.go:
```diff
@@ -38,6 +38,8 @@ type Config struct {
 	// torrents config
 	Torrents TorrentsConfig `yaml:"torrents,flow,omitempty"`
+	// upload config
+	Upload UploadConfig `json:"scraper" yaml:"upload,flow,omitempty"`
```
here json:"upload"

Kiloutre committed:
Update structs.go

Kiloutre committed:
Update upload.go

Kiloutre committed:
Update default_config.yml

Kiloutre committed:
Merge branch 'dev' into upload-multiple

Kiloutre committed:
Update upload_multiple.jet.html

Kiloutre committed:
Merge branch 'dev' into upload-multiple

Kiloutre committed:
Merge branch 'dev' into upload-multiple

Kiloutre committed:
Update upload.go

Kiloutre committed:
Update upload.jet.html

akuma06 committed:
Update template_test.go

Kiloutre committed:
this one is optional for anidex

Kiloutre committed:
Update default_config.yml

akuma06 committed:
Merge branch 'dev' into upload-multiple

akuma06 committed:
Possible improvement
As mentionned in my comment, it seems that having a checkbox already checked and disabled for logged in user + having the input text for apikey is a bit tedious.
Moreover, asking to check and fill the input text to upload in anidex when you are not logged is doing things twice. If you don't want to upload to anidex, don't fill the input.

So this commit push a new behaviour:
* If you are logged in and you have filled your anidex/nyaasi api key : the form should only show the checkbox asking if you want to upload to nyaasi.
* If you're not logged in or you don't have filled your api key in your profile settings: you only show the input text for the api key and not the checkbox. If someone wants to upload to nyaasi/anidex, he will just need to fill the input.

akuma06 committed:
This adds back the support of anonymous upload when logged in or not.
This works simply by checking if the user wants to upload as anon (when he checks the checkbox "upload as anonymous") or by checking the emptyness of apikey.
+ reverts the condition statements since it was needed to have apikey empty

Kiloutre commented:
> **If you're not logged in**: **you only show the input text for the api key** and not the checkbox. If someone wants to upload to nyaasi/anidex, he will just need to fill the input.

What for? The idea is to allow anyone to upload to anidex & nyaa.si very easily from pantsu, merely checking a checkbox when you don't want to have to go through the bother of getting an anidex & nyaa.si API key is massively better
Anidex's account creation doesn't work at all for me and nyaa.si are too incompetent to show your API key somewhere in your settings, yet alone having an apidoc. Well great, now this whole PR is completely useless for me

Furthermore you're making it so that absolutely every single anonymous upload gets uploaded to anidex nyaa.si, again what is the point? We should upload the torrent to those websites only if the user asks to do so, either with him giving his API key or by having our own used when asked, the only thing this will accomplish is get our pantsu account on anidex & nyaa.si account banned from those websites with the garbage we tend to get

I ask for a rollback

akuma06 commented:
Already reverted for non logged user if you checked the commit just after this one

Kiloutre commented:
Alright, the checkbox doesn't seem to work though?

akuma06 committed:
Forgot to save this file

akuma06 commented:
I think it would be better to have the upload to the three platform in a go routine or there will be timeout loading when someone do the multiupload.
Then you load the state with some AJAX or manual F5.

akuma06 committed:
This commit adds the go routines for each upload service (anidex, nyaasi, tosho).
New controller and url to check the upload status (/upload/status/:id).
If you add ?json at the end, it outputs a json format of the multi upload status.
To prevent memory overload, we only keep in memory the multiupload status for 5 minutes.
Moved multipleform struct from templates to upload utils.

Kiloutre committed:
Merge branch 'dev' into upload-multiple

Kiloutre committed:
fix form display on classic theme

akuma06 committed:
This commit adds:
* Javascript refreshing of the upload status
* Localization of the pages
* Fixed some bugs

This works, only need to get torrent file now.

akuma06 committed:
Added a new function to get the Path to the torrent file.

akuma06 committed:
Added a function to check that a torrent file exists and if not generate it on multiupload

akuma06 committed:
Enabled file upload through http post request to anidex

akuma06 committed:
Fixed bugs

akuma06 committed:
Modified the behaviour to generate torrent preivously. It is a synchrone function now. So we need to make it asynchronous in download.go

Now torrents file are always generated at upload time in a background processus with the correct trackers (needed ones + the ones provided by the user).
I made the anidex tracker as a needed one to allow multiupload even if the user hasn't included in his torrent file/magnet url.

akuma06 commented:
It works but it seems there is a problem with how anidex detects their tracker in our torrent file.
For example, I tested to send the torrent file generated through nyaa to anidex (with anidex tracker in it, I checked in my bitorrent client) and it gives me an error "Anidex tracker not included".
I tested through http://torrenteditor.com/ to check again that there was anidex tracker. It was. Then I downloaded from their website (without any changes) the torrent file and uploaded it to anidex. And it worked.
So I don't know what might be wrong. If someone can test it to see if there is the same problem

Kiloutre commented:
Maybe it has to do with tracker order?
Looking at anidex torrent files, anidex is always at the very start of the trackers. Furthermore in their listing page, any tracker beside anidex is omitted, so i wouldn't be surprised if they only checked the very first tracker to see if it's the anidex one

Kiloutre committed:
Merge branch 'dev' into upload-multiple

Kiloutre commented:
I'm getting a 404 error when trying to upload to other websites, the torrent is uploaded & torrent file generated but /upload/status/:id gives me a 404
I also get this error message in the console
`[GIN-debug] [WARNING] Headers were already written. Wanted to override status code 302 with 500`

Torrent generation also doesn't seem to work during a regular download.go call, it says it's generating but it's seemingly not

akuma06 commented:
Ok I will look into it when I have time :)

akuma06 committed:
Merge branch 'dev' into upload-multiple

akuma06 committed:
Moving deadtrackers list to default config
Support torrent file creation when upload without scraping
* Fix possible bug where no one is seeding a torrent yet (first time upload)
Add primary tracker in torrent file

akuma06 committed:
Fixing issue with boolean pointer for private torrent

akuma06 committed:
fixing index out of range for announcelist filtering
Fixing anidex tracker not found

akuma06 committed:
Fixing lang id error

akuma06 committed:
Anidex upload works

akuma06 committed:
Fix for multiupload with magnet

akuma06 commented:
So I've modified some parts to the torrent generation when uploading torrent file.
The download part works and the upload to anidex too.

Now we need to only fill in nyaasi/tosho form and request part.

akuma06 commented:
So I was looking to implement nyaasi upload but it seems that doing that would make less secure their credentials since nyaasi doesn't use an api key it seems.
I didn't find a documentation on how to use their api but I did find a python script (https://github.com/nyaadevs/nyaa/blob/master/utils/api_uploader_v2.py). It looks like you have to use a basic http authentification to use their api.
Beside the fact that asking the credentials from nyaasi to allow upload is not secure, I couldn't get their api to work with my Rest Client (same with their python file). It is always sending me this error: 
```
{
  "errors": {
    "recaptcha": [
      "The response parameter is missing."
    ]
  }
}
```

If someone can tell me if their python script is up to date and how their api works, that would be awesome!

akuma06 committed:
This adds TokyoTosho support. Since TokyoTosho API only support upload by URL (and not torrent file upload). We need to make the download url available without ddos protection.

Also TokyoTosho doesn't need a category select. Since it does have limited option, we can automatically convert our categories to the tokyotosho one.

akuma06 committed:
Removing the folder tosho and put categories.go directly in upload package.
Added category conversion for nyaasi too.
Added tests on category conversion.

akuma06 committed:
This should add nyaasi support.
Is not tested though.

akuma06 committed:
Updated defaut config with new nyaa.si behaviour?

akuma06 committed:
Forgot to commit this file

ewhal commented:
<3
